samples/face_reid/src/index_builder_client.py

```diff
@@ -0,0 +1,162 @@
+"""Index builder module."""
+import argparse
+import os
+import pathlib
+import shutil
+import time
+
+import cv2
+import hnswlib
+import numpy as np
+
+from samples.face_reid.utils import pack_person_id_img_n
+from savant.client import JpegSource, SinkBuilder, SourceBuilder
+from savant.utils.logging import get_logger, init_logging
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--zmq_src_socket', default=os.environ.get('ZMQ_SRC_ENDPOINT'))
+    parser.add_argument(
+        '--zmq_sink_socket', default=os.environ.get('ZMQ_SINK_ENDPOINT')
+    )
+    parser.add_argument('--source_id', default='gallery-img')
+    parser.add_argument('--index_dir', default='/index')
+    parser.add_argument('--gallery_dir', default='/gallery')
+    parser.add_argument('--shutdown_auth', default='shutdown')
+    # face recognition model parameters
+    parser.add_argument('--feature_namespace', default='adaface_ir50_webface4m_90fb74c')
+    parser.add_argument('--feature_name', default='feature')
+    parser.add_argument('--index_space', default='cosine')
+    # index_dim is set according to the reid model output dimensions
+    parser.add_argument('--index_dim', default=512, type=int)
+    # hnswlib index parameters
+    # for details refer to hnswlib doc
+    # https://github.com/nmslib/hnswlib#api-description
+    parser.add_argument('--index_max_elements', default=100, type=int)
+    parser.add_argument('--index_ef_construction', default=200, type=int)
+    parser.add_argument('--index_m', default=16, type=int)
+
+    return parser.parse_args()
+
+
+def main(args):
+    logger = get_logger('index_client')
+
+    logger.info('Starting Savant client...')
+
+    processed_gallery_dir = os.path.join(args.index_dir, 'processed_gallery')
+    index_file_path = os.path.join(args.index_dir, 'index.bin')
+
+    shutil.rmtree(processed_gallery_dir, ignore_errors=True)
+    os.makedirs(processed_gallery_dir)
+
+    # Build the source
+    source = SourceBuilder().with_socket(args.zmq_src_socket).build()
+
+    sink = SinkBuilder().with_socket(args.zmq_sink_socket).with_idle_timeout(10).build()
+
+    src_jpegs = [
+        JpegSource(args.source_id, str(img_path))
+        .with_aspect_ratio((16, 9))
+        .with_source_id_add_size_suffix(True)
+        for img_path in pathlib.Path(args.gallery_dir).glob('*.jpeg')
+    ]
+
+    source_ids = set()
+    for src_jpeg in src_jpegs:
+        source(src_jpeg, send_eos=False)
+        source_ids.add(src_jpeg.source_id)
+
+    for src_id in source_ids:
+        source.send_eos(src_id)
```

Comment:
This is done in order to send EOS only once for each source_id.

The only way to avoid SeqId discrepancy warnings from `savant_rs::message::validate_seq_iq` (both from the pipeline and the client) that I've found is to break up the images into several "streams", i.e. source_ids, in each of which all the frames have the same size, and to send EOS only once for each such "stream".

Reply:
@abramov-oleg does EOS after each image leads to the warning emition?